---
title: "Cell lines integration task"
author: "Ant√≥nio Sousa (e-mail: <aggode@utu.fi>) - Elo lab (<https://elolab.utu.fi>)"
date: "`r format(Sys.Date(), '%d/%m/%Y')`"
output: 
   html_document:
      toc: true 
      toc_float: true
      theme: united
      code_folding: hide
---

<br>

<br>

---

<br>

<br>

# Integration task

<br>

This single-cell RNA-seq integration task describes an example of cell lines integration: Jurkat versus HEK293T versus a 50:50 mixture of Jurkat:HEK293T cell lines. The data sets were downloaded from 10X genomics website (check the R script `01_create_datasets.R`): 

   + Jurkat (`jurkat`): 3,257 cells
   
   + HEK293T (`t293`): 2,854 cells
   
   + 50:50 Jurkat:HEK293T (`jurkat_t293_50:50`): 2,553 cells
      
<br>

The identity of the data set - `jurkat` or `t293` or `jurkat_t293_50:50` - for every cell was saved in the `Seurat` `meta.data` column variable `batch`. The ground-truth cell identities were also provided in the column variable `cell_type` but avoid checking them until the end of this notebook to make these analyses more interesting. 

The analyses performed in this notebook rely in the `Seurat` R package (v.`r packageVersion("Seurat")`). 

```{r setup, include=FALSE}
## Set up configuration for R chunks
knitr::opts_chunk$set(
	echo = TRUE,
	fig.align = "center",
	message = FALSE,
	warning = FALSE
)
if ( ! "klippy" %in% row.names(installed.packages()) ) remotes::install_github("rlesur/klippy")
```

```{r klippy, echo=FALSE}
## Configure the layout of icon to copy to clipboard the R chunk codes
klippy::klippy(position = c('top', 'right'))
```

Import the main packages used in this notebook: `Seurat` (v.`r packageVersion("Seurat")`), `SeuratWrappers` (v.`r packageVersion("SeuratWrappers")` - integration wrappers for Seurat), `dplyr` (v.`r packageVersion("dplyr")` - wrangling data), `patchwork` (v.`r packageVersion("patchwork")` - visualization), `scIntegrationMetrics` (v `r packageVersion("scIntegrationMetrics")` - compute LISI integration metrics).

```{r packages, message=FALSE, warning=FALSE}
## Import packages
library("dplyr") # data wrangling
library("Seurat") # scRNA-seq analysis
library("patchwork") # viz
library("SeuratWrappers") # integration wrappers
library("scIntegrationMetrics") # compute LISI integration metrics
```

Create output directories to save intermediate results, figures, tables and R objects. 

```{r dirs}
## Output directories
res.dir <- file.path("../results", "cell_lines_task", c("plots", "tables", "objects"))
for (folder in res.dir) if (!dir.exists(folder)) dir.create(path = folder, recursive = TRUE)
```

<br>

<br>

---

<br>

<br>

## (1) Import datasets 

_(5 min)_

>AIM: Import and explore the Seurat object data. 

<br>

Import the cell lines `Seurat` R object `jurkat.rds` located in the folder `data`.  

```{r import data}
# Import data
data.dir <- "../data"
seu <- readRDS(file = file.path(data.dir, "jurkat.rds"))
```

<br>

### Downsample dataset 

<br>

You may want to down sample this data set depending on the amount of RAM memory you have. The `jurkat` data set has 8,664 cells. 

```{r downsample data set}
## Downsample data set
downsample <- TRUE # replace to FALSE in case you don't want to down sample
prop.down <- 0.4 # proportion of cells to down sample per batch: 40% of the cells
if (downsample) {
  no.cells.batch <- ceiling(table(seu$batch) * 0.4) # CTRL = 1310 and STIM = 1491 
  cell.idx.batch <- split(x = colnames(seu), f = seu$batch) # split into a list the cell names per batch
  cell.idx.batch.down <- lapply(X = setNames(names(cell.idx.batch), names(cell.idx.batch)), FUN = function(x) {
    set.seed(123)
    sample(x = cell.idx.batch[[x]], size = no.cells.batch[[x]], replace = FALSE)
  }) # downsample each batch cell names 
  cell.idx.downsample <- do.call(c, cell.idx.batch.down) # join cell name labels from the two batches into one character vector
  seu <- subset(seu, cells = cell.idx.downsample)
}
gc()
cat("No. of cells downsampled for `jurkat` was:", table(seu$batch)[1], "\n")
cat("No. of cells downsampled for `jurkat_t293_50:50` was:", table(seu$batch)[2], "\n")
cat("No. of cells downsampled for `t293` was:", table(seu$batch)[3], "\n")
```

Explore quickly the `Seurat` `seu` object.

```{r explore Seurat object}
## Explore Seurat object
# Print Seurat object
seu

# Structure
str(seu)

# Check meta.data
head(seu@meta.data)

# Check how many cells per data set 
table(seu$batch)

# Check no. of genes 
nrow(seu)

# Check no. of cells 
ncol(seu)
```

<br>

<br>

---

<br>

<br>

## (2) Assess batch effect

<br>

### Joint dimred

_(7 min)_

>AIM: See how much the two data sets overlap each other in the low dimensional reductions. 

<br>

Run the standard `Seurat` upstream workflow to jointly compute a PCA and UMAP for the datasets: 

   1. `NormalizeData()`: log1p-normalization with a scaling factor of 10K
   
   2. `FindVariableFeatures()`: identification of 2K HVG
      
   3. `ScaleData()`: standardization of the 2K HVG
         
   4. `RunPCA()`: computation of a PCA with the 2K HVG standardized
   
   5. `RunUMAP()`: computation of a UMAP using the first `dims` of the previously computed PCA

```{r joint dimred analysis}
## Joint analysis

# Standard Seurat upstream workflow
seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu)
seu <- ScaleData(seu)
seu <- RunPCA(seu)
seu <- RunUMAP(seu, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")
```

Plot the PCA and UMAP side-by-side below. 

```{r plot joint dimreds, fig.width=10, fig.height=4}
## Plot jointly dimreds
pca.unint <- DimPlot(seu, reduction = "pca", group.by = "batch")
umap.unint <- DimPlot(seu, reduction = "umap.unintegrated", group.by = "batch")
pca.unint + umap.unint
```

<br>

### Celltype markers

_(5 min)_

>AIM: Check if cells from different datasets share well-known cell-specific markers. 

<br>

Plot below cell lines-specific markers: `CD3D` for jurkat and `XIST` for HEK193T. 

```{r joint celltype markers}
## Joint celltype markers

# List of jurkat and T293 cell lines
markers.plot <- list(
  "jurkat" = "CD3D", 
  "t293" = "XIST"
)

# Plot
jurkat.markers.unint.plot <- FeaturePlot(seu, features = markers.plot$jurkat, split.by = "batch", 
                                       max.cutoff = 3, cols = c("grey", "red"), 
                                       reduction = "umap.unintegrated", ncol = 4, 
                                       pt.size = 0.5)
t293.markers.unint.plot <- FeaturePlot(seu, features = markers.plot$t293, split.by = "batch", 
                                           max.cutoff = 3, cols = c("grey", "red"), 
                                           reduction = "umap.unintegrated", ncol = 4, 
                                           pt.size = 0.5)
```

```{r plot joint celltype markers, fig.width=12, fig.height=4}
## Plot jointly celltype markers

# Print 
jurkat.markers.unint.plot 
t293.markers.unint.plot
```

<br>

<br>

---

<br>

<br>

## (3) Integrate datasets

_(10 min)_

>AIM: Compare different integration methods. 

<br>

First, `split` the layers of data by batch before performing integration. Then, apply the standard `Seurat` workflow. Finally, call the function `IntegrateLayers()` to integrate the datasets. In this function you can specify the method you want to run by providing the integration method function. 

`Seurat` provides three methods: CCA (`CCAIntegration`), RPCA (`RPCAIntegration`) and Harmony (`HarmonyIntegration`). In addition, other methods can be called by using functions from `SeuratWrappers` such as: FastMNN (`FastMNNIntegration`) or scVI (`scVIIntegration`) among others. Harmony (from the `harmony` R package), FastMNN (from the `batchelor` R package) and scVI (python package installed with conda) need to be installed independently from `Seurat`. 

Run the R chunk code below to run the integration methods: CCA, RPCA, Harmony and FastMNN (you can try to run scVI if you've it installed in your system). Join the layers back after integration to project the integrated data onto UMAP. The UMAP highlights the `batch` and ground-truth `cell_type` labels. 

```{r integration}
## Perform integration

# Split layers for integration
seu[["RNA"]] <- split(x = seu[["RNA"]], f = seu$batch)

# Standard workflow
seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu)
seu <- ScaleData(seu)
seu <- RunPCA(seu)

# Integrate layers
int.methods <- c("CCA" = "CCAIntegration", "RPCA" = "RPCAIntegration", 
                 "Harmony" = "HarmonyIntegration", "FastMNN" = "FastMNNIntegration", 
                 "scVI" = "scVIIntegration")

for (m in names(int.methods)[1:4]) {
  cat("\nRunning integration method", m, "...\n")
  int.dimred <- paste0("integrated.", m)
  umap.dimred <- paste0("umap.", m)
  # Integration
  if (m=="scVI") {
          seu <- IntegrateLayers(object = seu, method = get(eval(substitute(int.methods[m]))), 
                         orig.reduction = "pca", 
                         new.reduction = int.dimred,
                         conda_env = "~/miniconda3/envs/scvi-env", # substitute this by your installation 
                         verbose = TRUE)
  } else {
      seu <- IntegrateLayers(object = seu, method = get(eval(substitute(int.methods[m]))), 
                         orig.reduction = "pca", 
                         new.reduction = int.dimred,
                         verbose = TRUE)
  }

}

# Re-join layers after integration
seu[["RNA"]] <- JoinLayers(seu[["RNA"]])

# Run UMAP for every integration method
int.umaps.plots <- list()
for (m in names(int.methods)[1:4]) {
  cat("\nRunning UMAP for", m, "integrated result...\n")
  int.dimred <- paste0("integrated.", m)
  umap.dimred <- paste0("umap.", m)
  seu <- RunUMAP(seu, dims = 1:30, reduction = int.dimred, reduction.name = umap.dimred)
  int.umaps.plots[[m]] <-  DimPlot(object = seu, reduction = umap.dimred, group.by = c("batch", "cell_type"), 
                                   combine = FALSE, label.size = 2)
}

# Save Seurat object
saveRDS(object = seu, file = file.path(res.dir[3], "seu_integrated.rds"))
```

<br>

<br>

---

<br>

<br>

## (4) Assess integration

_(15 min)_

>AIM: Assess integration qualitatively and quantitatively through dimensional reduction visualizations and LISI scores.  

<br>

### Qualitative viz

<br>

Plot the integrated embeddings below highlighting the `batch` and ground-truth `cell_type` labels. 

```{r assess qualitative integration - plot, fig.width=14, fig.height=18}
## Assess integration by printing the plots using the "batch" and "cell_type" (ground-truth) labels
wrap_plots(c(int.umaps.plots$CCA, int.umaps.plots$RPCA, int.umaps.plots$Harmony, int.umaps.plots$FastMNN),
           ncol = 2, byrow = TRUE)
```

<br>

<br>

### Quantitative metrics

<br>

Run the code below to compute the i/cLISI scores for every integrated embedding with the function `getIntegrationMetrics()` from the package `scIntegrationMetrics` (read more about the meaning of these metrics [here](https://github.com/carmonalab/scIntegrationMetrics)).

```{r assess quantitative integration}
## Assess quantitatively integration with scIntegrationMetrics

# Calculate metrics
int.mthds.names <- paste0("integrated.", names(int.methods)[1:4])
names(int.mthds.names) <- int.mthds.names
metrics <- list()
for (m in int.mthds.names) {
  key <- gsub("integrated.", "", m)
  cat("Computing i/cLISI metrics for integration method:", gsub("integrated.", "", key), "\n")
  metrics[[key]] <- getIntegrationMetrics(seu, meta.label = "cell_type", meta.batch = "batch",
                                          method.reduction = m, metrics = c("iLISI", "norm_iLISI", 
                                                                            #"CiLISI", "CiLISI_means", 
                                                                            "norm_cLISI", "norm_cLISI_means"))
}
# Join metrics
metrics <- as.data.frame(do.call(cbind, metrics))
```

Print the result below. 

```{r assess quantitative integration - table}
# Print table
knitr::kable(metrics)
```

<br>

<br>

---

<br>

<br>

#### R packages used and respective versions

<br>

```{r References, message=FALSE, warning=FALSE, paged.print=FALSE}
## R packages and versions used in these analyses
sessionInfo()
```

<br>

<br>

---

<br>

<br>
