# Integration

This folder serves as a repository for the course materials provided on the second day of the course _The Hitchhikerâ€™s Guide to scRNA-seq_ (08-12/07/2024, iMM, Lisbon, Portugal), which focuses on the integration of single-cell RNA sequencing data.

The course materials utilize minimalistic, publicly available scRNA-seq datasets to evaluate the performance of various integration methods across three tasks: (1) cross-tissue, (2) biological conditions, and (3) cell lines. The course addresses the challenges of integration, highlighting the strengths and weaknesses of different methods applied to these tasks. Additionally, a practical example of reference mapping is demonstrated by classifying and projecting PBMCs from two COVID-19 patients onto a previously annotated PBMC reference.

All the data used to start the analyses and the respective generated intermediate results have been deposited in Zenodo. The software, data and code required to reproduce the analyses are distributed through a docker image. More details about the course materials can be found in the website. 

The links for the resources used in the second day of this course appear below: 

   + Website: [https://elolab.github.io/Hitchhikers_Guide_scRNAseq_course](https://elolab.github.io/Hitchhikers_Guide_scRNAseq_course)

   + Zenodo: [https://zenodo.org/records/12620772](https://zenodo.org/records/12620772)

   + Docker image: [https://hub.docker.com/r/elolabfi/imm_scrnaseq_course](https://hub.docker.com/r/elolabfi/imm_scrnaseq_course)
