{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyN0MK/pncMeGYqoEkP+tnZb"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"m19nnevNHzbF"},"source":["# Using scanpy for scRNA-seq analysis"]},{"cell_type":"markdown","metadata":{"id":"ICWO4DpEN21e"},"source":["In this notebook we will be using scanpy.\n","\n","The notebook also includes some code to run pseudotime analysis, but this will be explored on a different tutorial in a later session."]},{"cell_type":"markdown","metadata":{"id":"TWcREeCiH7y4"},"source":["We start by installing some required packages"]},{"cell_type":"code","source":["!pip install scanpy scvelo anndata umap-learn matplotlib leidenalg"],"metadata":{"id":"w5AlSU_MZAvr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"DdY14PhMIArZ"},"source":["And we import those packages"]},{"cell_type":"code","metadata":{"id":"fnh_gqTi7R4N"},"source":["import scanpy as sc # package for manipulating single cell data\n","sc.settings.verbosity = 3 # getting more info from scanpy\n","import scvelo as scv # package based on scanpy to do advanced modelling of RNA velocity\n","\n","import pandas as pd # general data wranggling\n","import numpy as np # numeric data wranggling\n","\n","# setting the seed for RNG\n","import random\n","random.seed(10)\n","\n","# plotting\n","import matplotlib as mpl\n","import matplotlib.pyplot as plt\n","## magic to allow in-line figures\n","%matplotlib inline"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"MPAeBJSRJRE8"},"source":["Setting visualisation parameters for scvelo"]},{"cell_type":"code","metadata":{"id":"FHxdDZS_7Upd"},"source":["scv.settings.verbosity = 3  # show errors(0), warnings(1), info(2), hints(3)\n","scv.settings.presenter_view = True  # set max width size for presenter view\n","scv.settings.set_figure_params('scvelo')  # for beautified visualization"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"YYc851XeIMuC"},"source":["## Basic concepts about AnnData objects"]},{"cell_type":"markdown","metadata":{"id":"cpTlpsfaE0xD"},"source":["Obtain the data - we're using one of the datasets included in scVelo. This is a mouse pancreas development dataset from https://journals.biologists.com/dev/article/146/12/dev173849/19483/Comprehensive-single-cell-mRNA-profiling-reveals-a"]},{"cell_type":"code","metadata":{"id":"MR-Gi5l7E6C5"},"source":["adata = scv.datasets.pancreas()\n","adata"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"moUGYEpqPHu-"},"source":["adata.obs.head()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Axg6mvRsJxfl"},"source":["You can just see the number of rows and columns of an AnnData object (or any DataFrame) with `.shape`."]},{"cell_type":"code","metadata":{"id":"wkF7IW--JpOJ"},"source":["adata.shape"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"CUPrtLidFMC3"},"source":["`.obs` contains the cell metadata (i.e. annotations, etc.)."]},{"cell_type":"code","metadata":{"id":"xEJX_XEmFD6M"},"source":["adata.obs.head(n = 10)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"g_CtX-zuGAhi"},"source":["`.var` contains the gene metadata"]},{"cell_type":"code","metadata":{"id":"OHQ2gw1UF9TU"},"source":["adata.var.head(n = 10)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"EecsFBOUGK1q"},"source":["Each of these tables is implemented as a Pandas `DataFrame`. The row names are in `.index`, and the column names are in `.columns`. These, in turn, are numpy arrays, that can be accessed with `.values`"]},{"cell_type":"code","metadata":{"id":"GtjhhHsRGLFZ"},"source":["adata.var.index.values"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"U7NB0vXSGjdp"},"source":["adata.obs.columns.values"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"QYqS_rayGq0j"},"source":["You can count the number of elements in each category with `.value_counts()`"]},{"cell_type":"code","metadata":{"id":"RJbq7OcQGq-J"},"source":["adata.obs[\"clusters_coarse\"].value_counts()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"KfdS2Al-GrKQ"},"source":["And you can cross tabulate two columns using `.crosstab()` (a function in the pandas package)"]},{"cell_type":"code","metadata":{"id":"TH5eeWWBGrR0"},"source":["pd.crosstab(adata.obs[\"clusters_coarse\"], adata.obs[\"clusters\"])"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"NuDtm_SUs1Da"},"source":["Doubts about how to use a function?"]},{"cell_type":"code","metadata":{"id":"PdcRhXhhs4UB"},"source":["??sc.tl.pca"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"o27DfaS6KRAA"},"source":["## Basic analysis"]},{"cell_type":"markdown","metadata":{"id":"fii4Gt1MTKv4"},"source":["Adapted from: https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html"]},{"cell_type":"markdown","metadata":{"id":"R6Bw3tMBE9Q0"},"source":["Check the expression of the most expressed genes"]},{"cell_type":"code","metadata":{"id":"mTY4V1m2JE5t"},"source":["sc.pl.highest_expr_genes(adata, n_top=20)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"s44pQ28BNeYM"},"source":["Filtering cells and genes based on their expression level. By convention, cells are not considered as such with less than 100 different expressed genes (and usually we filter above that). Genes should also be present in a non-trivial portion of the data. We have 0.1% (or 3/~3000 cells) here."]},{"cell_type":"code","metadata":{"id":"0NLW4dSqJE3I"},"source":["sc.pp.filter_cells(adata, min_genes=200)\n","sc.pp.filter_genes(adata, min_cells=3)\n","adata"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"FY7TVEy5OTM-"},"source":["Now we're calculating some important QC metrics."]},{"cell_type":"code","metadata":{"id":"2R_0uirvJE8K"},"source":["adata.var['mt'] = adata.var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n","sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n","adata.obs.head(10)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"Qhvd5eHEJE-Y"},"source":["sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'],\n","             jitter=0.4, multi_panel=True)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"Ne3bDN47KNfS"},"source":["sc.pl.scatter(adata, x='total_counts', y='pct_counts_mt')\n","sc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"i4s-ycssP5su"},"source":["And we can now filter the data. We'll be removing some cells that have too many counts or MT% (even though with the latter I'm just being picky)"]},{"cell_type":"code","metadata":{"id":"6v3DeLEGKNhf"},"source":["adata = adata[adata.obs.total_counts < 16000, :]\n","adata = adata[adata.obs.pct_counts_mt < 2, :]"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"SjCBqIdpQHB9"},"source":["Now we can normalise and log-transform the data"]},{"cell_type":"code","metadata":{"id":"nFwb269AKNjk"},"source":["sc.pp.normalize_total(adata, target_sum=1e4)\n","sc.pp.log1p(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"mSktB5HGQNiK"},"source":["And then calculate the highly-variable genes"]},{"cell_type":"code","metadata":{"id":"HJnrO0CBKNl7"},"source":["sc.pp.highly_variable_genes(adata, min_mean=0.01, max_mean=3, min_disp=0.5)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"aIhh_gETQRlq"},"source":["We can use scanpy's plotting functions to examine gene variability and expression level"]},{"cell_type":"code","metadata":{"id":"ZU-dGFGHKNoD"},"source":["sc.pl.highly_variable_genes(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"hyOHBf_ZQA14"},"source":["scanpy doesn't automatically keep different slots for the unchanged data, even after filtering to use only HVG. So first we must save the normalised data into a \"raw\" slot, so we can later perform differential expression"]},{"cell_type":"code","metadata":{"id":"c8QR8pNTKNqa"},"source":["adata.raw = adata"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"FsYouNh7P-4e"},"source":["adata = adata[:, adata.var.highly_variable]"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"pujrt1ZEQ0s6"},"source":["And now we filter the main matrix to only contain the HVG"]},{"cell_type":"code","metadata":{"id":"4Iv6tnF8P-7F"},"source":["sc.pp.regress_out(adata, ['total_counts'])"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"b-daXHuzRG2X"},"source":["We scale those genes"]},{"cell_type":"code","metadata":{"id":"gp03XpkBP--c"},"source":["sc.pp.scale(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OiKflstdRJTF"},"source":["And run a PCA"]},{"cell_type":"code","metadata":{"id":"X0kloMzjP_AN"},"source":["sc.tl.pca(adata, svd_solver='arpack')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"nctMyavMQeG-"},"source":["sc.pl.pca(adata, color=['Por', 'Col18a1'])"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"DJ7pJXNjXhYg"},"source":["adata"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"x8JSKniARNq9"},"source":["How much variance does each PC explain?"]},{"cell_type":"code","metadata":{"id":"UxUmzPgwQeJb"},"source":["sc.pl.pca_variance_ratio(adata, log=True)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"BhbTlmZ5RSF1"},"source":["We can now calculate the neighbourhood graph for the cells in PC space"]},{"cell_type":"code","metadata":{"id":"X9DCFeKJQeL4"},"source":["sc.pp.neighbors(adata, n_neighbors=10, n_pcs=25)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"dHNy_BbLRlFw"},"source":["And get a UMAP"]},{"cell_type":"code","metadata":{"id":"W8pgYQ5GQeOS"},"source":["sc.tl.umap(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"F5JN9jPxQeQ0"},"source":["sc.pl.umap(adata, color=['Col18a1'], use_raw=False)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"XsVjoZTpSLgN"},"source":["We also cluster the data"]},{"cell_type":"code","metadata":{"id":"gdmivAnzQeTS"},"source":["sc.tl.leiden(adata, resolution = 0.4)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"G70_TLtgRajI"},"source":["sc.pl.umap(adata, color=['leiden'])"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"3ZLMytLISbOv"},"source":["Now we can calculate marker genes for all these clusters! We will also be plotting the top 25 genes for each of them."]},{"cell_type":"code","metadata":{"id":"IPvP2j6IRhpk"},"source":["sc.tl.rank_genes_groups(adata, 'leiden', method='wilcoxon', use_raw = True)\n","sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"UAgTxcF7Sy9a"},"source":["We can also print the top 5 genes for all clusters as a table"]},{"cell_type":"code","metadata":{"id":"sj6_WA0DRalc"},"source":["pd.DataFrame(adata.uns['rank_genes_groups']['names']).head(5)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OIIHQmiMS4Al"},"source":["And we can include more information by also showing the p-value for the DE test"]},{"cell_type":"code","metadata":{"id":"JUnJx_KzSIlU"},"source":["result = adata.uns['rank_genes_groups']\n","groups = result['names'].dtype.names\n","pd.DataFrame(\n","    {group + '_' + key[:1]: result[key][group]\n","    for group in groups for key in ['names', 'pvals']}).head(5)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Sefe_sU8TFd9"},"source":["How does cluster 0 compare against the rest for its top 8 marker genes?"]},{"cell_type":"code","metadata":{"id":"MRg-Dh40SuSb"},"source":["sc.pl.rank_genes_groups_violin(adata, groups='0', n_genes=8)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"BBe2xSKoSQlS"},"source":["We can also do 1 vs 1 comparisons"]},{"cell_type":"code","metadata":{"id":"FmSgbOZLSInp"},"source":["sc.tl.rank_genes_groups(adata, 'leiden', groups=['0'], reference='1', method='wilcoxon')\n","sc.pl.rank_genes_groups(adata, groups=['0'], n_genes=20)\n","sc.pl.rank_genes_groups_violin(adata, groups='0', n_genes=8)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"NWQOkvMSSIp8"},"source":["sc.tl.rank_genes_groups(adata, 'leiden', groups=['1'], reference='0', method='wilcoxon')\n","sc.pl.rank_genes_groups(adata, groups=['1'], n_genes=20)\n","sc.pl.rank_genes_groups_violin(adata, groups='1', n_genes=8)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"q1SeTKbiT_i_"},"source":["Lets define a set of marker genes here for further plotting"]},{"cell_type":"code","metadata":{"id":"3ZPcMH9JT8af"},"source":["marker_genes = ['Dlk1', 'Spp1', 'Krt7', 'Top2a', 'Vtn', 'Sox9', 'Isl1', 'Cldn4', 'Pyy']"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"GA_RmKspSIs1"},"source":["sc.pl.violin(adata, marker_genes, groupby='leiden')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"-6WmiApiSyK5"},"source":["sc.pl.umap(adata, color=['leiden']+marker_genes, legend_loc='on data', title='', frameon=False, save='.pdf')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ZZKsfE3yW6GT"},"source":["To get an overview of cell populations, we can look at heatmaps/dotplots/violin matrices"]},{"cell_type":"code","metadata":{"id":"xD2SNCpsSyNd"},"source":["sc.pl.dotplot(adata, marker_genes, groupby='clusters');"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"5T5BzqTqS6Ue"},"source":["sc.pl.stacked_violin(adata, marker_genes, groupby='clusters');"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"5992p4CkSaDC"},"source":["And this is how you save the data (but you don't have to do it here)"]},{"cell_type":"code","metadata":{"id":"8ZL9WpKtRan3"},"source":["# adata.write(\"path/to/output/adata.h5ad\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"SN1jvKuuR4Yy"},"source":["## Pseudotime analysis"]},{"cell_type":"markdown","metadata":{"id":"B-EE0j0nDPrm"},"source":["### PAGA - Partition-based graph abstraction (see https://scanpy-tutorials.readthedocs.io/en/latest/paga-paul15.html)"]},{"cell_type":"markdown","metadata":{"id":"RQb-73tDXCy9"},"source":["PAGA can help us understand the relationship between clusters, and lead to a cluster-based pseudotime ordering of the data"]},{"cell_type":"code","metadata":{"id":"xifm4cp9RaqW"},"source":["sc.tl.paga(adata, groups='leiden')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"Yo1Tqr_bCiZC"},"source":["sc.pl.paga(adata, color=['leiden']+marker_genes)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"8hJnWYOzXpep"},"source":["And now we can get a new graph-based projection for the data"]},{"cell_type":"code","metadata":{"id":"-w2grHUICibj"},"source":["sc.tl.draw_graph(adata, init_pos='paga')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"lxgSBWXWCieF"},"source":["sc.pl.draw_graph(adata, color=['leiden'] + marker_genes, legend_loc='on data')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"GS-DE8FSDcBZ"},"source":["### Diffusion pseudotime"]},{"cell_type":"markdown","metadata":{"id":"2Ud13dAgZiif"},"source":["To run a diffusion-based pseudotime algorithm, we have first to define a root cell"]},{"cell_type":"code","metadata":{"id":"0ah5dYuMDzcy"},"source":["adata.uns['iroot'] = np.flatnonzero(adata.obs['leiden']=='16')[0]"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"m32jkl4tZoIj"},"source":["Now we get the diffusion pseudotime"]},{"cell_type":"code","metadata":{"id":"VNWlKXULDNTy"},"source":["sc.tl.dpt(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"iH9CY0eiZreu"},"source":["And plot it"]},{"cell_type":"code","metadata":{"id":"_ojbJgGcDNWN"},"source":["sc.pl.draw_graph(adata, color=['leiden', 'dpt_pseudotime'], legend_loc='on data')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"sjqBYPVylRn7"},"source":["sc.pl.paga(adata, color=['leiden'])"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"9SL_DVVHsiXN"},"source":["And we can plot change in gene expression along pseudotime"]},{"cell_type":"code","metadata":{"id":"3CwvwW9OlGQA"},"source":["scv.pl.scatter(adata, x='dpt_pseudotime', y=marker_genes[0], frameon=False, use_raw = None)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OV3Bv8SEsmRk"},"source":["The follwing code can be found in the tutorial https://scanpy-tutorials.readthedocs.io/en/latest/paga-paul15.html. Should show plotting for each defined branch, but unfortunately it doesn't work"]},{"cell_type":"code","metadata":{"id":"8F_XhlZ2Ras3"},"source":["paths = [('beta', [17, 12, 7, 13, 10,0,4,9,5,3,2,6,11]),\n","         ('alpha', [17, 12, 7, 13, 10,0,4,9,5,3,2,16,1]),\n","         ('delta', [17, 12, 7, 13, 10,0,4,9,5,3,2,15,8])]\n","\n","_, axs = plt.subplots(ncols=3, figsize=(6, 2.5), gridspec_kw={'wspace': 0.05, 'left': 0.12})\n","plt.subplots_adjust(left=0.05, right=0.98, top=0.82, bottom=0.2)\n","\n","for ipath, (descr, path) in enumerate(paths):\n","  _, data = sc.pl.paga_path(\n","      adata, path, marker_genes,\n","      show_node_names=False,\n","      ax=axs[ipath],\n","      ytick_fontsize=12,\n","      left_margin=0.15,\n","      n_avg=50,\n","      annotations=['dpt_pseudotime'],\n","      #show_yticks=True if ipath==0 else False,\n","      show_colorbar=True,\n","      color_map='Greys',\n","      groups_key='leiden',\n","      color_maps_annotations={'dpt_pseudotime': 'viridis'},\n","      title='{} path'.format(descr),\n","      return_data=True,\n","      show=False)\n","\n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"kLgrXLKYR7PE"},"source":["## scVelo analysis for transcriptional dynamics"]},{"cell_type":"markdown","metadata":{"id":"THVJzwyBIc89"},"source":["The dataset we're using already includes a quantification for spliced and unspliced transcripts"]},{"cell_type":"code","metadata":{"id":"2yt4BLb0QeVq"},"source":["scv.utils.show_proportions(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ve6mDV51tAlr"},"source":["Further, we need the first and second order moments (basically mean and uncentered variance) computed among nearest neighbors in PCA space. First order is needed for deterministic velocity estimation, while stochastic estimation also requires second order moments."]},{"cell_type":"code","metadata":{"id":"Gy-XlWvkQeYI"},"source":["scv.pp.moments(adata, n_pcs=30, n_neighbors=30)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"n1i4seZQtBRs"},"source":["The gene-specific velocities are obtained by fitting a ratio between precursor (unspliced) and mature (spliced) mRNA abundances that well explains the steady states (constant transcriptional state) and then computing how the observed abundances deviate from what is expected in steady state\n","\n"]},{"cell_type":"code","metadata":{"id":"jr508dJMQeaj"},"source":["scv.tl.velocity(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"BXyvLkWutB0P"},"source":["Now we compute a velocity graph for low-dimensional embedding\n","\n"]},{"cell_type":"code","metadata":{"id":"JrP_v_E7P_Ct"},"source":["scv.tl.velocity_graph(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"db4cBin1w3O3"},"source":["sc.pl.umap(adata, color=['leiden'])"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"esWRXShyP_FQ"},"source":["scv.pl.velocity_embedding_stream(adata, basis='umap',  color = \"leiden\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"JA4R1DacP_Hj"},"source":["scv.pl.velocity_embedding(adata, basis='umap', arrow_length=2, arrow_size=1.5, dpi=150)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zeGB1Nh1tGjW"},"source":["Dynamic modelling is done to get the full transcriptional dynamics (see https://scvelo.readthedocs.io/DynamicalModeling/). We can get transcription, splicing and degradation rates (not covered here)\n"]},{"cell_type":"code","metadata":{"id":"nAs_9swSP_KB"},"source":["scv.tl.recover_dynamics(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"ffBxvLjoP_Nj"},"source":["scv.tl.velocity(adata, mode='dynamical')\n","scv.tl.velocity_graph(adata)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"oegP_HRmuO1I"},"source":["We can also obtain a latent pseudotime estimate"]},{"cell_type":"code","metadata":{"id":"2sRZUR5BP_WH"},"source":["scv.tl.latent_time(adata)\n","scv.pl.scatter(adata, color='latent_time', color_map='gnuplot', size=80, colorbar=True)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"oi7Hkvl8KNsc"},"source":["top_genes = adata.var['fit_likelihood'].sort_values(ascending=False).index[:300]\n","scv.pl.heatmap(adata, var_names=top_genes, tkey='latent_time', n_convolve=100, col_color='clusters')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"393LD8SLKNuk"},"source":["scv.pl.scatter(adata, x='latent_time', y=marker_genes[0], frameon=False)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"9zHLmc_fKNwx"},"source":["scv.pl.velocity_embedding_stream(adata, basis='umap')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"yTNTM-g4KNy_"},"source":[],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"92suCG8HKN1U"},"source":[],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"dEaUzSFjKN3m"},"source":[],"execution_count":null,"outputs":[]}]}